lib/buffer.js

```diff
@@ -24,82 +24,148 @@ function createPool() {
   }
 
-  if (this.length > kMaxLength) {
-    throw new RangeError('Attempt to allocate Buffer larger than maximum ' +
-                         'size: 0x' + kMaxLength.toString(16) + ' bytes');
+  // Slightly less common case.
+  if (typeof(arg) === 'string') {
+    fromString(this, arg, arguments.length > 1 ? arguments[1] : 'utf8');
+    return;
   }
 
-  if (this.length < 0)
-    this.length = 0;
-  else
-    this.length >>>= 0;  // Coerce to uint32.
+  // Unusual.
+  fromObject(this, arg);
+}
 
-  this.parent = undefined;
-  if (this.length <= (Buffer.poolSize >>> 1) && this.length > 0) {
-    if (this.length > poolSize - poolOffset)
-      createPool();
-    this.parent = sliceOnto(allocPool,
-                            this,
-                            poolOffset,
-                            poolOffset + this.length);
-    poolOffset += this.length;
-  } else {
-    alloc(this, this.length);
-  }
+function fromNumber(that, length) {
+  allocate(that, length < 0 ? 0 : checked(length) >>> 0);
+}
 
-  if (typeof subject === 'number') {
-    return;
+function fromString(that, string, encoding) {
+  if (typeof(encoding) !== 'string' || encoding === '')
+    encoding = 'utf8';
+
+  // Assumption: byteLength() return value is always < kMaxLength.
+  var length = byteLength(string, encoding) >>> 0;
+  allocate(that, length);
+
+  var actual = that.write(string, encoding) >>> 0;
+  if (actual !== length) {
+    // Fix up for truncated base64 input.  Don't bother returning
+    // the unused two or three bytes to the pool.
+    that.length = actual;
+    truncate(that, actual);
   }
+}
 
-  if (typeof subject === 'string') {
-    // In the case of base64 it's possible that the size of the buffer
-    // allocated was slightly too large. In this case we need to rewrite
-    // the length to the actual length written.
-    var len = this.write(subject, encoding);
-    // Buffer was truncated after decode, realloc internal ExternalArray
-    if (len !== this.length) {
-      var prevLen = this.length;
-      this.length = len;
-      truncate(this, this.length);
-      // Only need to readjust the poolOffset if the allocation is a slice.
-      if (this.parent != undefined)
-        poolOffset -= (prevLen - len);
-    }
+function fromObject(that, object) {
+  if (object instanceof Buffer)
+    return fromBuffer(that, object);
 
-  } else if (subject instanceof Buffer) {
-    subject.copy(this, 0, 0, this.length);
+  if (Array.isArray(object))
+    return fromArray(that, object);
+
+  if (object == null)
+    throw new TypeError('must start with number, buffer, array or string');
 
-  } else if (typeof subject.length === 'number' || Array.isArray(subject)) {
-    // Really crappy way to handle Uint8Arrays, but V8 doesn't give a simple
-    // way to access the data from the C++ API.
-    for (var i = 0; i < this.length; i++)
-      this[i] = subject[i];
+  if (object.buffer instanceof ArrayBuffer)
+    return fromTypedArray(that, object);
```

Comment:
That would only work for Int8Array / Uint8Array, I think.  Other typed arrays need the `element & 255` masking for compatibility reasons (which I think is silly design mistake, but that aside.)

Idle musing: the overhead of calling into C++ probably isn't worth it for small arrays.  I don't care enough about typed arrays yet to go find the crossover point, though.